### PR TITLE
Run the supervisor in a studio

### PR DESCRIPTION
**Pull request by mwrock**
This starts an empty supervisor inside of an interactive studio. The supervisor is started in the background and is terminated when the studio exits. A convenience function is added to the studio profile `tail_sup` (very open to better names) which runs `tail -f` on the log file that the supervisor output is piped to.

Signed-off-by: Matt Wrock <matt@mattwrock.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by mwrock** [comment]
This is all working in a linux studio. I want to double check docker and add documentation before merging.

**Comment by mwrock** [comment]
This fixes #2131

File: components/studio/bin/hab-studio.sh [hunk 1]
```diff
@@ -679,11 +679,25 @@ enter_studio() {
...
+  $bb chroot "$HAB_STUDIO_ROOT" \
     $studio_env_command -i $env $studio_enter_command $*
 }
 
+# **Internal** Run when an entered studio exits.
+stop_supervisor() {
+  lock_file="$HAB_STUDIO_ROOT/hab/sup/default/LOCK"
+
+  if [ -f $lock_file ]; then
+    $bb kill -9 $($bb cat $lock_file)
```
Comment (reset) [inline components/studio/bin/hab-studio.sh]: Should we try to let the supervisor gracefully exit instead of forcing it with `-9` immediately?

File: components/studio/libexec/hab-studio-type-default.sh [hunk 2]
```diff
@@ -84,6 +86,12 @@ exec $HAB_ROOT_PATH/bin/hab pkg exec core/hab-plan-build hab-plan-build \$*
 EOF
   $bb chmod $v 755 $HAB_STUDIO_ROOT$HAB_ROOT_PATH/bin/build
 
+  $bb cat <<TAIL_SUP > $HAB_STUDIO_ROOT$HAB_ROOT_PATH/bin/tail_sup
```
Comment (reset) [inline components/studio/libexec/hab-studio-type-default.sh]: What about just simply calling it `log` or `logs` like how we have the `build` command built in?

Comment (reset) [inline components/studio/libexec/hab-studio-type-default.sh]: By the way, holy shit was this a good idea

**Comment by fnichol** [comment]
I still need to try this out, but I want to avoid us running software by default inside a Studio. This would incur more software being downloaded and installed on each new Studio instance and isn't relevant for post-build failures (when a `hab pkg build` fails and you want to `hab studio enter` to trace an error).

Could we add an explicit command inside the Studio to do a similar thing so the user knows a Supervisor is being executed? I don't like the idea of other active processes running in the build environment when a package build is being executed. At least not without asking for it. Otherwise maybe use an opt-in flag to do the same thing on a `hab studio enter`? I'd still much rather see a `run-sup` or equivalent command inside the Studio though.

File: components/studio/libexec/hab-studio-type-default.sh [hunk 3]
```diff
@@ -6,6 +6,7 @@ studio_build_environment=
 studio_build_command="record \${1:-} $HAB_ROOT_PATH/bin/build"
 studio_run_environment=
 studio_run_command="$HAB_ROOT_PATH/bin/hab pkg exec core/hab-backline bash -l"
+studio_supervisor_start_command="$HAB_ROOT_PATH/bin/hab sup run > /hab/sup/log.out &"
```
Comment (reset) [inline components/studio/libexec/hab-studio-type-default.sh]: You should place this in `/hab/sup/default/out.log` - the supervisor that we're starting is "named" default. If somebody starts a separate supervisor they might want to consider outputting that to `/hab/sup/<name>/out.log`

**Comment by reset** [comment]
@mwrock super quick and clean work man!

gif-keyboard-6463566812690874366

File: components/studio/bin/hab-studio.sh (hunk 1 above)
Comment (mwrock) [inline components/studio/bin/hab-studio.sh]: fur sure. reflecting on this, I have no idea why I used `-9`

File: components/studio/libexec/hab-studio-type-default.sh (hunk 2 above)
Comment (mwrock) [inline components/studio/libexec/hab-studio-type-default.sh]: To me `log<s>` feels vague. Its unclear if it is build log or supervisor logs. In a studio, my assumption would be build logs.

**Comment by mwrock** [comment]
I can see your point @fnichol . I see this as being helpful to the mac development workflow where a studio is the only place you can run a supervisor. Honestly I think that if we ask the user t ostart the supervisor with a command, we might as well just guide them to run `hab sup run`.

Thinking about this more, I wonder if it makes sense to totally rethink the api around starting services. Having a `hab start` command that does 2 things based on the state of the supervisor is kind of weird. Its also weird having `hab start` and `hab load` doing nearly the same thing if the supervisor is running. I know this might sound extreme but what if we just eliminate `hab start` or deprecate it. I'm just gonna stop myself here but worth thinking about.

File: components/studio/libexec/hab-studio-type-default.sh (hunk 2 above)
Comment (smith) [inline components/studio/libexec/hab-studio-type-default.sh]: I think this is fine for now but somebody should open an issue to replace this with `hab sup logs -f`.

Since this is for studio use, I'm not concerned about the growth of this log file, but if habitat starts producing logs anywhere but stdout/stderr or the system journal, it has a responsibility to rotate them.

Comment (mwrock) [inline components/studio/libexec/hab-studio-type-default.sh]: Yeah I think that ideally we would have a `hab sup start -d` and the `-d` would indicate daemonized and it would emit logs in a sane way to a known location with rotation. Then `hab sup log` would consume those logs. I'm tempted just to do that here, but should just leave it for a separate pr.

Comment (reset) [inline components/studio/libexec/hab-studio-type-default.sh]: This can't become a command to `hab-sup` because we don't actually log to file by convention. We're just redirecting output here in the studio to provide this particular functionality - so it would make sense that this stay a command of the studio

**Comment by reset** [comment]
@fnichol we don't need to run the supervisor if the studio is initiated by `hab pkg build`, but I think it definitely makes sense when using an interactive studio with `hab studio enter`.

The most common concern that we've heard after releasing the multi-tenant supervisor is the confusion around `hab start` having what is viewed as a "dual purpose". I think we actually got this UX completely correct, but we need to make small changes to our docs and studio to have it feel natural to people. In most cases, you will be using `hab start` and `hab stop` to do one of two things:

# Start/Stop a transient service in an interactive studio so you can test your package or use `--config-from` to test iterations on your hooks, etc
# Modify a permanent service's desired state in a production environment to start or stop

The studio would be the place that we would push people to would accomplish #1. We can additionally ask somebody to run `hab sup run > /path/to/log &` and then tail it, but the goal of this PR was to automate that and abstract away tailing logs to a single, easy to remember command, just like `build` from within the studio

**Comment by reset** [comment]
Per the offline discussion with @mwrock, @fnichol, and myself, we're gonna make a few changes before merging this in

* Add an environment variable (perhaps `BG_SUP=false` or something along these lines?) which can be set in a `.studiorc` to disable automatic starting of a backgrounded supervisor in an interactive studio
* Ensure that we don't automatically start a backgrounded supervisor in a non-interactive studio (I think this might already be done?)
* Add some output when entering the interactive studio stating if the backgrounded service is started or stopped. If it's stopped, mention the env var that disabled it. If it's started, we should introduce the `slog` command and tell people how to `start`/`stop` services, if there's room mention `--config-from`, and maybe link to some docs if that's helpful
* Rename the `tail_sup` command to `slog` (short for supervisor log)

**Comment by jonsmorrow** [comment]
This is awesome and I dig the pre merge changes.

**Comment by mwrock** [comment]
```
vagrant@vagrant:/vagrant/components/studio$ hab studio enter
   hab-studio: Creating Studio at /hab/studios/vagrant--components--studio (default)
   hab-studio: Importing core secret origin key
» Importing origin key from standard input
★ Imported secret origin key core-20160810182414.
   hab-studio: Entering Studio at /hab/studios/vagrant--components--studio (default)
   hab-studio: Exported: HAB_ORIGIN=core

** The Habitat Supervisor has been started in the background.
** Use 'hab sup load' and 'hab sup unload' to load and unload services.
** Use the 'slog' command to stream the supervisor log.
** Adding '$NO_BG_SUP=1' to your .studiorc file will disable the background supervisor.

[1][default:/src:0]# slog
hab-sup(MR): Butterfly Member ID 1c9c5d7601274140b318fc8d1d8946d9
hab-sup(MR): Starting butterfly on 0.0.0.0:9638
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
connecting to tcp://localhost:10001
connecting to tcp://localhost:10011
connecting to tcp://localhost:10021
♥
[2][default:/src:130]# exit
logout
vagrant@vagrant:/vagrant/components/studio$
```

Is the heart an accident?...I don't think so.

**Comment by smith** [comment]
> `** Use 'hab sup load' and 'hab sup unload' to load and unload services.`

Should we prefer `start` and `stop` in a studio, since they allow `--config-from` and you're more likely going to want transient services here?

> `** Adding '$NO_BG_SUP=1' to your .studiorc file will disable the background supervisor.`

No `$` here.

**Comment by mwrock** [comment]
thanks @smith fixed both comments

**Comment by reset** [comment]
gif-keyboard-7821617314214773675

@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/224489188).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/224489188).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/224489188).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265
